codebasin/config.py

```diff
@@ -270,6 +306,225 @@ def load_codebase(config, rootdir):
     return codebase
 
 
+_importcfg = None
+def load_importcfg():
+    """
+    Load the import configuration file, if it exists.
+    """
+    global _importcfg
+    _importcfg = collections.defaultdict(list)
+    path = ".cbi/importcfg.json"
+    if os.path.exists(path):
+        log.info(f"Found import configuration file at {path}")
+        with open(path, "r") as f:
+            try:
+                _importcfg_json = json.load(f)
+                for compiler in _importcfg_json["compilers"]:
+                    _importcfg[compiler["name"]] = compiler["options"]
+            except BaseException:
+                log.error("importcfg file failed validation")
+
+
+class Compiler(object):
+    """
+    Represents the behavior of a specific compiler, including:
+    - The number of passes it performs.
+    - Implicitly defined macros, which may be flag-dependent.
+    """
+
+    def __init__(self, args):
+        self.name = os.path.basename(args[0])
+        self.args = args
+        self.passes = set(["default"])
+
+        # Check for any user-defined compiler behavior.
+        # Currently, users can only override default defines.
+        if _importcfg is None:
+            load_importcfg()
+        self.defines = extract_defines(_importcfg[self.name])
+
+    def get_passes(self):
+        return self.passes.copy()
+
+    def get_defines(self, pass_):
+        return self.defines.copy()
+
+    def get_include_paths(self, pass_):
+        return []
+
+    def get_include_files(self, pass_):
+        return []
+
+    def has_implicit_behavior(self, pass_):
+        return (self.get_defines(pass_) or
+                self.get_include_paths(pass_) or
+                self.get_include_files(pass_))
+
+    def get_configuration(self, pass_):
+        defines = self.get_defines(pass_)
+        include_paths = self.get_include_paths(pass_)
+        include_files = self.get_include_files(pass_)
+        return {"defines": defines,
+                "include_paths": include_paths,
+                "include_files": include_files}
+
+
+class ClangCompiler(Compiler):
+    """
+    Represents the behavior of Clang-based compilers.
+    """
+
+    device_passes = [
+        "spir64",
+        "x86_64",
+        "spir64_x86_64",
+        "spir64_gen",
+        "spir64_fpga"
+    ]
+
+    def __init__(self, args):
+        super().__init__(args)
+
+        self.sycl = False
+        self.omp = False
+        sycl_targets = []
+
+        for arg in args:
```

Comment:
Personally, I would refactor this loop/behavior as a `Compiler` method as opposed to subclasses...

Reply:
Could you give an example of what you mean here?

The reason I set this up as subclasses is because I convinced myself there was a hierarchical relationship between the compilers.  Clang implements almost every Gnu flag for compatibility reasons, and compilers like icpx extend the base Clang behavior with vendor-specific flags.

I'm not opposed to another solution, I'm just not sure what you're suggesting.

Reply:
Sorry, now that I read my message again I did realize it didn't make much sense...

I meant moving the `for arg in args` logic out of individual `__init__` methods for each subclass, and just have the logic defined in another function (e.g. `Compiler.process_args`).

But based on your other reply, there's probably enough differences between compiler behaviors that this isn't a straightforward refactor... Let's ignore for now

Reply:
Oh, I see what you mean.  I'm happy to ignore this for now, too -- but I agree we should try to refactor a lot of this code.